design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,434 @@
+
+Progress will be updated by volume snapshotter in VolumePluginBackup CR which is specific to that backup operation.
+
+### Progress of restore operation handled by volume snapshotter
+
+Progress will be updated by volume snapshotter in VolumePluginRestore CR which is specific to that restore operation.
+
+## Detailed Design
+
+### Approach 1
+
+Existing `Snapshot` Go struct from `volume` package have most of the details related to backup operation performed by volumesnapshotters.
+This struct also gets backed up to backup location. But, this struct doesn't get synced on other clusters at regular intervals.
+It will be synced only during restore operation, and velero CLI shows few of its contents.
+
+At a high level, in this approach, this struct will be converted to a CR by adding new fields (related to Progress tracking) to it.
+Instead of backing up of Go struct, CRs will be backed up to backup location, and it gets synced into other cluster by backupSyncController running in that cluster.
+
+#### VolumePluginBackup CR
+
+There is one addition to volume.SnapshotSpec, i.e., ProviderName. Below is the updated one:
+
+```
+type SnapshotSpec struct {
+	// BackupName is the name of the Velero backup this snapshot
+	// is associated with.
+	BackupName string `json:"backupName"`
+
+	// BackupUID is the UID of the Velero backup this snapshot
+	// is associated with.
+	BackupUID string `json:"backupUID"`
+
+	// Location is the name of the VolumeSnapshotLocation where this snapshot is stored.
+	Location string `json:"location"`
+
+	// PersistentVolumeName is the Kubernetes name for the volume.
+	PersistentVolumeName string `json:persistentVolumeName`
+
+	// ProviderVolumeID is the provider's ID for the volume.
+	ProviderVolumeID string `json:"providerVolumeID"`
+
+	// Provider is the Provider field given in VolumeSnapshotLocation
+	Provider string `json:"provider"`
+
+	// VolumeType is the type of the disk/volume in the cloud provider
+	// API.
+	VolumeType string `json:"volumeType"`
+
+	// VolumeAZ is the where the volume is provisioned
+	// in the cloud provider.
+	VolumeAZ string `json:"volumeAZ,omitempty"`
+
+	// VolumeIOPS is the optional value of provisioned IOPS for the
+	// disk/volume in the cloud provider API.
+	VolumeIOPS *int64 `json:"volumeIOPS,omitempty"`
+}
+```
+
+Few fields (except first two) are added to volume.SnapshotStatusSpec. Below is the updated one:
+```
+type SnapshotStatus struct {
+	// ProviderSnapshotID is the ID of the snapshot taken in the cloud
+	// provider API of this volume.
+	ProviderSnapshotID string `json:"providerSnapshotID,omitempty"`
+
+	// Phase is the current state of the VolumeSnapshot.
+	Phase SnapshotPhase `json:"phase,omitempty"`
+
+	// PluginSpecific are a map of key-value pairs that plugin want to provide
+	// to user to identify plugin properties related to this backup
+	// +optional
+	PluginSpecific map[string]string `json:"pluginSpecific,omitempty"`
```

Comment:
as this is supposed to be properties specific to Backup/restore operation, thinking of changing to
`BackupProperties` and `RestoreProperties`..
does that make sense

Reply:
Would the properties be the same on backup and restore? Are they copied from the backup CR to the restore CR?

Reply:
Need not be @nrb .. Backup related and Restore related ones are different.
As mentioned in one of the comments, one Backup property can be its parent snapshot name, in case of OpenEBS as it supports incremental backups.

Reply:
`BackupProperties` or `BackupInfo` would work.